readme.md
#AKEA

These are some short notes on how to use the AKEA package and some background on the project.

##What is AKEA

The name AKEA origins from the name of our home super cluster Laniakea, which agein was named by words from the Hawaiian language. There lani stands for heaven (which is obviously not our business) and akea for spacious, immeasurable, which ist more what we are goping for. I came up with an acronym which might work as well: AKEA actually konws 'em all. Why this might not be an exxaggeration will be explained later.

AKEA konsists of several tools, which are all centered arround the usage of the AKEA database. The tools I will go into detail are called:
- QRAKEA
- DBAKEA
- ...

##So let's start with the AKEA concept
A classical file system uses adresses to identify the position of a file. Just two examples are NTFS and ZFS, while the latter already offers an 128 bit address space, which is virtually infinite, if you consider a computer fitted onto earth of any size.
AKEA goes down a different road, which allows to identify any file by a hash of itself. In short this is some kind of "fingerprint" which can be generated from a file and does not change as long as the file does not change. Hashing a file with a "good" hash algorithm to generate such a "fingerprint" has some properties which are quite interesting for data handling. 
The hash ("fingerprint") is potentially unique. Uniqueness is enshured by the pseudo random behaviour of the algorithm. Nevertheless, there is the slightest chance, that two different datasets get the same fingerprint which is called collision. This is the reason why AKEA does not only use 128 bit hash keys. Mathematically we are talking about the birthday problem (read about it on [wikipedia](https://en.wikipedia.org/wiki/Birthday_problem)).
Which allows us to calculate the probability of a collions for a given hash length l, and amount of files n:
$$
p(n)\approx 1-e^{{-n^2}/{2\cdot l}}
$$
So let us assume we are using 42 byte hashes (336 bit) which gives us 

$l=2^{336}\approx 1.4\cdot 10^{101}$ 

and have 

$n = 2^{128} \approx 3.4 \cdot 10^{38}$ 

files, we get:

$p(3.4 \cdot 10^{38})\approx 1-e^{{-(3.4 \cdot 10^{38})^2}/{(2\cdot 1.4\cdot 10^{101})}}\approx 4.13 \cdot 10^{-25}$

as the probability that a collision will occure for one (two) of these files.

When you do the math for [all atoms on this planet](https://education.jlab.org/qa/mathatom_05.html) 

$n \approx 1.33 \cdot 10^{50}$

you will get away with a remaingin probability of 

$p(1.33 \cdot 10^{50})\approx 6 \%$

which is still worth a bet that AKEA could actually know 'em all. Or at least would need all of earth's atoms to be able to store enough hashes to reach considerable collision chances.

##Why use QRAKEA?

QRAKEA is a fairly simple tool or rather script. It takes a file calculates the SHA512 hash of it and converts the first 42 bytes of it into a QR code. But why bother generating QR codes of files which you already have? Well in my case I was looking for a possibility to document and later find files I generate while e.g. doing a measurement or running a simulation script. These files typically do not change any more and are perfectly idetentifiable by their hash "fingerprint". The QR code than can be printed and placed in your labbook, analog or digital, to get you to your file later on. You even may transfer and rename all folders and filenames, without having to worry how to find them again.

##What is DBAKEA for?
When you now have the AKEA hashes it is desireable to find your files again. QRAKEA is only able to help you to verify that this is actually the file you once created. But what happens if you do not have a clou which file it is or were it went? Well then grab DBAKEA give your best bet were to find the file and hash a complete folder and its sub folders. If you then "show" the QR-code to DBAKE it will tell you if and were to find the file(s).

docs/changelog.rst
changelog
================================

tests/__init__.py
#import unittest
#from . import test_akea
#from . import test_dbakea
#from . import test_qrakea
#
#if __name__ == "__main__":
#    unittest.main()

tests/test_akea.py
import unittest
from akea import akea

ref_binary_digest               = ('./akea_test_dummies/test_dummie1.txt', '_sha512-336_',b'\xae#vm\x87\t,\xd2\x9dd\x8c\xd7L\r\xb7\xf2%\x9du\x15\xb0c<\xc3\x12\xb1Z\x81\x9e\xb3\x96\x94J,\x81\xfc\x8d\x92\x86\xcc\xae\x18')
ref_binary_digest_with_folder   = ('./akea_test_dummies/', '_sha512-336_',[b'\xae#vm\x87\t,\xd2\x9dd\x8c\xd7L\r\xb7\xf2%\x9du\x15\xb0c<\xc3\x12\xb1Z\x81\x9e\xb3\x96\x94J,\x81\xfc\x8d\x92\x86\xcc\xae\x18', b'\xae#vm'])
ref_hex_digest                  = ('./akea_test_dummies/test_dummie1.txt', '_sha512-336x_','ae23766d87092cd29d648cd74c0db7f2259d7515b0633cc312b15a819eb396944a2c81fc8d9286ccae18')
ref_hex_digest_with_folder      = ('./akea_test_dummies/', '_sha512-336x_',['ae23766d87092cd29d648cd74c0db7f2259d7515b0633cc312b15a819eb396944a2c81fc8d9286ccae18', 'ae23766d'])



class AKEAtest(unittest.TestCase):

    def testAKEA_hash_and_file_functions(self):
        """Test the AKEA hashing functions.

        Test all AKEA basic functions by using some test files.

        Args:
            self: object

        Returns:
            None
        """
        self.assertEqual(akea.binary_digest('./akea_test_dummies/test_dummie1.txt'), ref_binary_digest)
        self.assertEqual(akea.hex_digest('./akea_test_dummies/test_dummie1.txt'), ref_hex_digest)
        self.assertEqual(akea.binary_digest('./akea_test_dummies/'), ref_binary_digest_with_folder)
        self.assertEqual(akea.hex_digest('./akea_test_dummies/'), ref_hex_digest_with_folder)



if __name__ == "__main__":
    unittest.main()

tests/test_dbakea.py
import unittest
from akea import dbakea
import configparser

class DBAKEAtest(unittest.TestCase):

    def testDBAKEA_hash_and_file_functions(self):
        """Test the DBAKEA hashing functions.

        Test  DBAKEA basic functions by creating and adding to dbakea base and
        than check backwards by reading the qrakea file.

        Args:
            self: object

        Returns:
            None
        """
        config = configparser.ConfigParser()
        config.read('test_akea.conf')
        dbakea.append_to_db('./akea_test_dummies/', config)     #E:/Programmieren/AKEA/tests
        dbakea_entry = dbakea.check_qr(config, './akea_test_dummies/test_dummie1.txt.qrakea')
        self.assertEqual(dbakea_entry, './akea_test_dummies\\test_dummie1.txt')

if __name__ == "__main__":
    unittest.main()

akea/dbakea.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 11 21:09:20 2017

@author: Raoul-Amadeus Lorbeer
"""

import os           #get the filename separated

import sqlite3

from akea import akea
from akea import qrakea

#import configparser

def add_file_list(dbakea_cursor, dbakea, dirname, skip_sub_dirs=False):
    """append given file to AKEA database

    append given file to AKEA database.

    Args:
        dbakea_cursor (cursor object): cursor object to edit given database
        dbakea (database object): database object with opened file connection
        dirname (str): path to file or folder
        skip_sub_dirs (bool): subdirs are note added if true
    """
    print('Adding folder: ' + dirname)

    tag = 'none'
    version = 1.0
    file_list = []
    if os.path.isdir(dirname):
        file_list = akea.__non_qrakea_file_list(dirname, not skip_sub_dirs)
    else:
        file_list.append(dirname)

    #if skip_sub_dirs:
        #names = file_list#@todo: edit names to ommit subdirs!
        #return

    for file_name_i in file_list:    #add all files in list to database
        current_path = file_name_i      #obsolete?: os.path.join(dirname, file_name_i)
        try:
            data = akea.binary_digest(current_path)
            print(current_path)
            dbakea_cursor.execute("INSERT INTO files VALUES (?,?,'empty','empty',?,?)", \
                                      (current_path, memoryview(data.digest), tag, version))
        except sqlite3.Error:
            dbakea_cursor.execute('SELECT * FROM files WHERE AKEA=?', (memoryview(data.digest),))
            db_filepaths = dbakea_cursor.fetchone()[0]
            if db_filepaths.find(current_path) < 0:
                dbakea_cursor.execute('UPDATE files SET path = ? WHERE AKEA = ?', \
                                      ((db_filepaths + '; '+current_path), \
                                       memoryview(data.digest)))
            #else:
                #print("file known")
        except IOError:
            print("Could not access: " + current_path)
    dbakea.commit()

def append_to_db(path, config):
    """append given file to AKEA database

    append given file to AKEA database.

    Args:
        path (str): path to file or folder
        config (configuration parser = dictionary): system configuration
    """
    print(config['DBAKEA']['dbpath'])
    dbakea = sqlite3.connect(str(config['DBAKEA']['dbpath']))
    dbakea_cursor = dbakea.cursor()

    # Create table
    #check if table already exists!!
    try:
        dbakea_cursor.execute('''CREATE TABLE files
        (path text, AKEA blob, parent text, child text, TAG text, Version real, PRIMARY KEY (AKEA))
        ''')
    except sqlite3.Error:
        print("Table found")

    add_file_list(dbakea_cursor, dbakea, path)


def check_qr(config, image_file=''):
    """Search for file with QR code.

    Search for file with QR code from web-cam.

    Args:
        config (configuration parser = dictionary): system configuration

    Returns:
        filepath (str): path to file
    """
    the_data = 0
    if image_file != '':
        the_data = qrakea.get_qr_image(image_file)

    if the_data == 0:
        the_data = qrakea.get_qr_webcam(config)

    #search database entry
    dbakea = sqlite3.connect(config['DBAKEA']['dbpath'])
    dbakea_cursor = dbakea.cursor()
    dbakea_cursor.execute('SELECT * FROM files WHERE AKEA=?', (memoryview(the_data),))

    #get path / link
    db_filepaths = dbakea_cursor.fetchone()[0]
    print(db_filepaths)

    #close all
    dbakea.close()

    return str(db_filepaths)

akea/qrakea.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 11 21:09:20 2017

Always use the first 42 bytes of the SHA512 digest.
For one it is faster then the SHA256 and second you can allways "upgrade"
to full 512 bit key without loss of compatibility for old keys!

@author: Raoul-Amadeus Lorbeer
"""
#https://pypi.python.org/pypi/qrcode

from os import path#, system           #get the filename separated

import textwrap
import binascii
import qrcode       #convert to an image

from pyzbar.pyzbar import decode
import cv2
#import binascii
from PIL import Image
from PIL import ImageDraw
from PIL import ImageFont

#import configparser

def generate_qr_code(data):
    """Generate a QR code from given data.

    Generate a QR code from given data.

    Args:
        akea_digest (named tuple): digested tuple of data containing
                                    (path (str containing the original path),
                                    type (str identifying the AKEA spec),
                                    AKEA digest (AKEA HASH in binary)) compare akea_digest
    Returns:
        img (PIL image object): containing QR code
    """
    #some initial values
    border_width = 160 #px
    qr_box_size = 10
    qr_border = border_width/qr_box_size

    #qr setup
    qr_obj = qrcode.QRCode(
        version=1,
        error_correction=qrcode.constants.ERROR_CORRECT_Q,
        box_size=qr_box_size,
        border=qr_border,
    )

    #Add data to QR code
    qr_obj.add_data(path.basename(data.path))
    if path.basename(data.path) == '':
        qr_obj.add_data(data.path)
    qr_obj.add_data(data.type)
    qr_obj.add_data(data.digest)
    qr_obj.make(fit=True)

    img = qr_obj.make_image()

    #refurbish borders
    # The crop rectangle, as a (left, upper, right, lower)-tuple.
    img = img.crop((qr_box_size*(qr_border-4), qr_box_size*(qr_border-4), \
                    img.size[0]-qr_box_size*(qr_border-4), img.size[1]))

    text = data.path
    if (len(data.digest[0]) > 1) & (len(data.digest) >= 50):
        text += '\n Warning! Some files may have been skipped!'

    __add_text_to_img(img, text, offset=(qr_box_size*4, img.size[1] - border_width))

    return img

def __add_text_to_img(img, qr_text, offset):
    """Draw text into image.

    Draw text into image at given offset position.

    Args:
        img (PIL image object): image object containing image
        qr_text (str): text
        offset (tuple): position of text insertion

    Returns:
        img (PIL image object): image object containing image with text
    """

    draw = ImageDraw.Draw(img)
    textsize = 24
    text_font = ImageFont.truetype("pix12pt.ttf", size=textsize)

    #@todo: I have no idea why it is not 2*offset[0]
    file_lines = textwrap.wrap(qr_text, width=int(((img.size[0]-2.75*offset[0])/(textsize))))
    wraped_file_name = ''

    for line in file_lines:
        wraped_file_name += line
        wraped_file_name += '\n'

    #prevent crash on empty last line https://github.com/python-pillow/Pillow/issues/2614
    wraped_file_name = wraped_file_name[:-1]

    #@toto: make font work again!   , font = text_font)
    draw.multiline_text(offset, wraped_file_name, fill='black', font=text_font)

    return img

def get_qr_webcam(config):   #https://pypi.python.org/pypi/pyzbar/
    """Collect QR code from Web-cam.

    Collect QR code from Web-cam.

    Args:
        config (configuration dictionary): setup information

    Returns:
        QRAKEA (binary): QRAKEA hash
    """
    qr_data = ''

    #start with vide capture dev 0 (standard web-cam)
    cap = cv2.VideoCapture(int(config['QRAKEA']['device_number']))

    #Set image dimensions
    cap.set(5, int(config['QRAKEA']['fps']))
    cap.set(3, int(config['QRAKEA']['x_res']))
    cap.set(4, int(config['QRAKEA']['y_res']))

    while True:
        # get frame
        return_val, frame = cap.read()

        # convert image data for pyzbar
        img_gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        pil_object = Image.fromarray(img_gray)
        image = decode(pil_object)

        # display the resulting frame (convenience) and mirror (flip) makes it easier to place
        cv2.imshow('Please show the QRAKEA code to the camera. Hit space to exit.', cv2.flip(img_gray, 1))

        #extract results
        if len(image) > 0:
            qr_data = str(image[0].data)

        #abort scanning if enter was hit or QR-code was found
        if (cv2.waitKey(10) & 0xFF == ord(' ')) | (qr_data != ''):
            break        #nothing to scan!

    #clean up window
    cv2.destroyAllWindows()
    cap.release()

    #extract hex hash from QRcode-string
    hash_start = qr_data.rfind("_sha512-336x_")+13
    file_hash = qr_data[hash_start:-1]
    the_data = binascii.unhexlify(file_hash)

    return the_data

def get_qr_image(image_path):   #https://pypi.python.org/pypi/pyzbar/
    """Collect QR code from image.

    Collect QR code from image.

    Args:
        image_path (str): path to image

    Returns:
        QRAKEA (binary): QRAKEA hash
    """
    qr_data = ''
    #start with vide capture dev 0 (standard webcam)

    try:
        pil_object = Image.open(image_path)
    except OSError:# as e:
        #print("Could not open image.\n")
        return 0

    image = decode(pil_object)
    # extract results
    if len(image) > 0:
        qr_data = str(image[0].data)

    #extract hex hash from QRcode-string
    hash_start = qr_data.rfind("_sha512-336x_")+13	#need to create some meeningful variable to get rid of this hard coded 13

    if hash_start < 13:
        return 0            #no AKEA tag found

    file_hash = qr_data[hash_start:-1]
    the_data = binascii.unhexlify(file_hash)

    return the_data


exec/gen_qrakea.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 11 21:09:20 2017

Always use the first 42 bytes of the SHA512 digest.
For one it is faster then the SHA256 and second you can allways "upgrade"
to full 512 bit key without loss of compatibility for old keys!

@author: Raoul-Amadeus Lorbeer
"""
import os          #get the filename separated
import configparser
import sys

from win32api import SetFileAttributes
from win32con import FILE_ATTRIBUTE_READONLY

sys.path.append("..")
from akea import akea, qrakea

##_________________________________________________________##

if __name__ == '__main__':

    config = configparser.ConfigParser()
    config.read('akea.conf')

    print('Number of arguments:', len(sys.argv), 'arguments.')
    print('Argument List:', str(sys.argv))
    #read a file:
    file_path = ''#"E:\\Programmieren\\AKEA\\depricated\\DBASH\\test\\test0011.bmp"

    to_print = 0    #don't print if necessarry
    do_folder = 0   #check if it is a whole folder

    if len(sys.argv) > 1:     #go through arguments passed to script
        command_list = enumerate(sys.argv[1:])
        for command in command_list:            #all but first argument
            if str(command).find('-') < 0:        #this should be the file!
                file_path = str(command)
                #make the file read only since we do not intend to modify them any more!
                SetFileAttributes(file_path, FILE_ATTRIBUTE_READONLY)
                data = akea.hex_digest(file_path)    #data = akea_digest(file_path)
                print(data)
                img = qrakea.generate_qr_code(data)
                img.save(file_path+'.qrakea', format='PNG', dpi=(600, 600))
                img.show()
            if str(command).find('-print') >= 0:  #to screen or to printer?
                file_path = str(next(command_list)[1])
                #make the file read only since we do not intend to modify them any more!
                SetFileAttributes(file_path, FILE_ATTRIBUTE_READONLY)
                data = akea.hex_digest(file_path)    #data = akea_digest(file_path)
                img = qrakea.generate_qr_code(data)
                img.save(file_path+'.qrakea', format='PNG', dpi=(600, 600))
                os.startfile(file_path + '.qrakea', 'print')
                #system(r'"C:\Program Files\IrfanView\i_view64.exe" ' + file_path + r'.png /print')


exec/access_dbakea.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 11 21:09:20 2017

Always use the first 42 bytes of the SHA512 digest.
For one it is faster then the SHA256 and second you can allways "upgrade"
to full 512 bit key without loss of compatibility for old keys!

@author: Raoul-Amadeus Lorbeer
"""
import configparser
import os           #get the filename separated
import sys
sys.path.append("..")
from akea import dbakea

def get_path(config, the_image):
       #open file
    try:  #if there is a file path attached to the command
        return dbakea.check_qr(config, the_image)
    except StopIteration:  #if not do it old school
        return dbakea.check_qr(config)

##Script starts here
##_________________________________________________________##

if __name__ == '__main__':

    print('Number of arguments:', len(sys.argv), 'arguments.')
    print('Argument List:', str(sys.argv))
    #read a file:
    #file_path = 'E:\Programmieren\QR\DBakea\\testB\\'

    config = configparser.ConfigParser()
    config.read('akea.conf')

    if len(sys.argv) > 1:     #go through arguments passed to script
        command_list = enumerate(sys.argv[1:])
        for command in command_list:
            if str(command).find('-') < 0:
                dbakea.append_to_db(command, config)
                break
            if str(command).find('-pathqr') >= 0:
                #open folder
                os.startfile("\"" + os.path.dirname(str.split(get_path(config, \
                                 next(command_list)[1]), ';')[0]) + "\"")
                break
            if str(command).find('-openqr') >= 0:
                #open file
                os.startfile("\"" + os.path.abspath(str.split(get_path(config, \
                                 next(command_list)[1]), ';')[0]) + "\"")
                break


akea/akea.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 11 21:09:20 2017

Always use the first 42 bytes of the SHA512 digest.
For one it is faster then the SHA256 and second you can always "upgrade" to full 512 bit key
without loss of compatibility for old keys!

@author: Raoul-Amadeus Lorbeer
"""
#https://pypi.python.org/pypi/qrcode
#https://docs.python.org/2/library/hashlib.html

import glob

#import copy
import collections  #use tuples in function returns without messing everything up!

import hashlib      #create some hashes

from os import path

def hex_digest(file_path):
    """Generate an AKEA hex digest

    Generate an AKEA hex digest as specified in the AKEA specification.

    Args:
        file_path (str): Path to specified file

    Returns:
        akea_digest (named tuple): digested tuple of data containing
        (path (str containing the original path),
        type (str identifying the AKEA spec), AKEA digest (AKEA HASH in hex))
    """
    #first check if we have a folder
    if path.isdir(file_path):         #no file only a folder
        akea = __hex_folder_digest(file_path)
    else:       #take first 42 bytes ->84 hex values of digest
        akea = hashlib.sha512(__get_file_data(file_path)).hexdigest()[0:84]
    return collections.namedtuple('akea_digest', ('path', 'type', 'digest')) \
                                        (file_path, '_sha512-336x_', akea)

def __hex_folder_digest(folder_path, max_files=50):
    """Generate an AKEA hex digest tuple

    Generate an AKEA hex digest tuple starting wit 42 byte and then continuing with
    4 byte segments as specified in the AKEA specification.

    Args:
        folder_path (str): Path to specified folder
        max_files (int): number of files which may be included in the hash tuple

    Returns:
        akea hex digest (tuple): digested tuple of data containing hex digests for
        up to 100 files in one folder
    """
    akea = []

    file_list = __non_qrakea_file_list(folder_path)

    if file_list == []:
        return akea

    #take first 42 bytes ->84 hex values of digest
    akea.append(hashlib.sha512(__get_file_data(file_list[0])).hexdigest()[0:84])

     #choose step size to assure <50 files to represent the whole set
    for file in file_list[:max(1,int((len(file_list)/max_files))):]:
        #take first 4 bytes ->8 hex values of digest
        akea.append(hashlib.sha512(__get_file_data(file)).hexdigest()[0:8])

    return akea

def binary_digest(file_path, max_files=100):
    """Generate an AKEA binary digest

    Generate an AKEA binary digest as specified in the AKEA specification.

    Args:
        file_path (str): Path to specified file
        max_files (int): number of files which may be included in the hash tuple

    Returns:
        akea_digest (named tuple): digested tuple of data containing
        (path (str containing the original path), type (str identifying the AKEA spec),
        AKEA digest (AKEA HASH in binary))
    """
    #first check if we have a folder
    if path.isdir(file_path):         #no file only a folder
        akea = __binary_folder_digest(file_path, max_files)
    else:       #take first 42 bytes of digest
        akea = hashlib.sha512(__get_file_data(file_path)).digest()[0:42]
    return collections.namedtuple('akea_digest', ('path', 'type', 'digest')) \
                                                (file_path, '_sha512-336_', akea)

def __binary_folder_digest(folder_path, max_files):
    """Generate an AKEA binary digest tuple

    Generate an AKEA binary digest tuple starting wit 42 byte and then continuing
    with 4 byte segments as specified in the AKEA specification.

    Args:
        folder_path (str): Path to specified folder
        max_files (int): number of files which may be included in the hash tuple

    Returns:
        akea binary digest (tuple): digested tuple of data containing hex digests
        for up to 100 files in one folder
    """
    akea = []

    file_list = __non_qrakea_file_list(folder_path)

    if file_list == []:
        return akea

    #take first 42 bytes
    akea.append(hashlib.sha512(__get_file_data(file_list[0])).digest()[0:42])

    #choose step size to assure <100 files to represent the whole set
    for file in file_list[:max(1,int((len(file_list)/max_files))):]:
        akea.append(hashlib.sha512(__get_file_data(file)).digest()[0:4])     #take first 4 bytes

    return akea

def __get_file_data(file_path):
    """Extract data from file

    Used to get all data included in the file at the specified path.

    Args:
        file_path (str): Path to specified file
        max_files (int): number of files which may be included in the hash tuple

    Returns:
        binary: data stored in specified file
    """

    file = open(file_path, 'rb')
    file_data = file.read()
    file.close()
    return file_data

def __non_qrakea_file_list(folder_path, subfolders=False):
    """Generate file list with or without sub-folders without "*.qrakea".

    Generate file list with or without sub-folders without "*.qrakea".

    Args:
        folder_path (str): path to be searched in

    Returns:
        filepath (list of str): paths to files aond folders in folder_path
    """
    if subfolders:
        file_list = [file for file in glob.glob(path.join(folder_path, r'**\*.*'), recursive=True)
                     if not path.splitext(file)[1] == '.qrakea']
    else:
        file_list = [file for file in glob.glob(path.join(folder_path, '*.*'))
                     if not path.splitext(file)[1] == '.qrakea']

    return file_list
